### PR TITLE
implement dashboard api for recommended files

### PR DESCRIPTION
**Pull request by icewind1991**
Implement the dashboard api from https://github.com/nextcloud/server/pull/26430 and https://github.com/nextcloud/server/pull/33658

requires:

- [x] https://github.com/nextcloud/server/pull/33658

### CODE REVIEW

**Comment by tobiasKaminsky** [comment]
Thanks for this first version.
I tested it and it works so far!

Some questions/remarks
- icon_class is not needed
- icon_url must always be a real url
- button name and url is missing (this is used e.g. at "recent activity" or "github notifications")

**Comment by tobiasKaminsky** [comment]
Bug: iconUrl must always be set, as clients cannot know via API which type it is, they cannot fall back to their placeholder.
This means that also for a folder / empty text file correctly tinted placeholder must be returned.

**Comment by PVince81** [comment]
linter is also not happy here

File: lib/Dashboard/RecommendationWidget.php [hunk 1]
```diff
@@ -68,4 +88,23 @@ public function load(): void {
...
+	public function getItems(string $userId, ?string $since = null, int $limit = 7): array {
+		$user = $this->userManager->get($userId);
+		if (!$user) {
+			return [];
+		}
+		$recommendations = $this->recommendationService->getRecommendations($user, $limit);
+
+		return array_map(function(IRecommendation $recommendation) {
+			$url = $this->urlGenerator->linkToRouteAbsolute('files.viewcontroller.showFile', ['fileid' => $recommendation->getNode()->getId()]);
+			$icon = $this->urlGenerator->linkToRouteAbsolute('core.Preview.getPreviewByFileId', [
```
Comment (PVince81) [inline lib/Dashboard/RecommendationWidget.php:101]: @tobiasKaminsky do you mean we should pre-check if a preview exists, and if yes, return the preview URL and if not return the mime type icon URL instead ?

**Comment by PVince81** [comment]
@tobiasKaminsky do you happen to have a curl call handy to call this API ? (so that someone else can take over and continue without needing the full setup)

File: lib/Dashboard/RecommendationWidget.php (hunk 1 above)
Comment (tobiasKaminsky) [reply lib/Dashboard/RecommendationWidget.php:101]: Yes, clients always need a real icon, may it be svg or png.

**Comment by tobiasKaminsky** [comment]
CURL:
```
curl --request GET \
  --url 'https://server.com/ocs/v2.php/apps/dashboard/api/v1/widget-items?format=json&limit=14&widgets%5B%5D=recommendations' \
  --header 'Authorization: Basic 123123' \
  --header 'OCS-APIRequest: true' \
  --header 'format: json'
 ```

File: lib/Dashboard/RecommendationWidget.php (hunk 1 above)
Comment (CarlSchwan) [reply lib/Dashboard/RecommendationWidget.php:101]: At least in Qt/QML, this is not required as you can easily detect if the image loaded and add a fallback (e.g. mimetype icon). See e.g.  https://invent.kde.org/frameworks/kirigami/-/blob/master/src/controls/Avatar.qml#L235

Comment (tobiasKaminsky) [reply lib/Dashboard/RecommendationWidget.php:101]: But we do not have any info about the mimetype.
And sending mimetype along side, and then have clients deteminate it and show correct placeholder is way more work than directly sending the correct image.
(widgets on Android / iOS are limited in opposite to our full apps, so we cannot do too much calculation/work there)

**Comment by ArtificialOwl** [comment]
It seems the `IRecommendation` already have a value [regarding the availability of a preview](https://github.com/nextcloud/recommendations/blob/a166ec886737f238e8cc2e1c110facbf3e729283/lib/Service/RecommendationService.php#L74-L76)

@PVince81 Need to clarify if this is enough

Example of result of the ocs request:

```
{
  "ocs": {
    "meta": {
      "status": "ok",
      "statuscode": 200,
      "message": "OK"
    },
    "data": {
      "recommendations": [
        {
          "subtitle": "",
          "title": "Talk",
          "link": "http://nc25.local/index.php/f/411",
          "iconUrl": "http://nc25.local/index.php/apps/theming/img/core/filetypes/folder.svg?v=0",
          "sinceId": "1663763213"
        },
        {
          "subtitle": "",
          "title": "Library.jpg",
          "link": "http://nc25.local/index.php/f/406",
          "iconUrl": "http://nc25.local/index.php/core/preview?x=256&y=256&fileId=406&c=e6abee908319368e8bbdf39ea20f770e",
          "sinceId": "1663763212"
        },
[...]
```

**Review by PVince81 — APPROVED** [review]
:+1: awesome

**Comment by PVince81** [comment]
@tobiasKaminsky can you review if this solution is now suitable for clients ? thanks

**Comment by tobiasKaminsky** [comment]
This works @ArtificialOwl :clap:

**Comment by ChristophWurst** [comment]
There was no backport @PVince81

**Pull request merged**